worlds/overcooked2/Logic.py

```diff
@@ -0,0 +1,3899 @@
+
+
+def meets_minimum_sphere_one_requirements(
+    levels: Dict[int, Overcooked2GenericLevel],
+) -> bool:
+
+    # 1-1, 2-1, and 4-1 are garunteed to be accessible on
+    # the overworld without requiring a ramp or additional stars
+    sphere_one = [1, 7, 19]
+
+    # 1-2, 2-2, 3-1 and 5-1 are almost always the next thing unlocked
+    sphere_twoish = [2, 8, 13, 25]
+
+    # Peek the logic for sphere one and see how many are possible
+    # with no items
+    sphere_one_count = 0
+    for level_id in sphere_one:
+        if (is_completable_no_items(levels[level_id])):
+            sphere_one_count += 1
+
+    sphere_twoish_count = 0
+    for level_id in sphere_twoish:
+        if (is_completable_no_items(levels[level_id])):
+            sphere_twoish_count += 1
+
+    return sphere_one_count >= 2 and \
+        sphere_twoish_count >= 2 and \
+        sphere_one_count + sphere_twoish_count >= 6
+
+
+def is_completable_no_items(level: Overcooked2GenericLevel) -> bool:
+    one_star_logic = level_logic[level.shortname][0]
+    (exclusive, additive) = one_star_logic
+
+    # print(f"\n{level.shortname}: {exclusive} / {additive}")
+
+    return len(exclusive) == 0 and len(additive) == 0
+
+
+# If key missing, doesn't require a ramp to access (or the logic is handled by a preceeding level)
+#
+# If empty, a ramp is required to access, but the ramp button is garunteed accessible
+#
+# If populated, a ramp is required to access and the button requires all levels in the
+# list to be compelted before it can be pressed
+#
+ramp_logic = {
+    "1-5": [],
+    "2-2": [],
+    "3-1": [],
+    "5-2": [],
+    "6-1": [],
+    "6-2": ["5-1"],  # 5-1 spawns blue button, blue button gets you to red button
+    "Kevin-1": [],
+    "Kevin-7": ["5-1"],  # 5-1 spawns blue button,
+                         # press blue button,
+                         # climb blue ramp,
+                         # jump the gap,
+                         # climb wood ramps
+    "Kevin-8": ["5-1", "6-2"],  # Same as above, but 6-2 spawns the ramp to K8
+}
+
+horde_logic = {  # Additive
+    ("Coin Purse", 0.7),
+    ("Calmer Unbread", 0.35),
+    ("Progressive Dash", 0.2),
+    ("Progressive Throw/Catch", 0.15),
+    ("Sharp Knife", 0.15),
+    ("Dish Scrubber", 0.125),
+    ("Burn Leniency", 0.1),
+    ("Spare Plate", 0.075),
+    ("Clean Dishes", 0.025),
+}
+
+# Level 1 - dict keyed by friendly level names
+# Level 2 - tuple with 3 elements, one for each star requirement
+# Level 3 - tuple with 2 elements, one for exclusive requirements and one for additive requirements
+# Level 4 (exclusive) - set of item name strings of items which MUST be in the inventory to allow logical completion
+# Level 4 (additive)  - list of tuples containing item name and item weight where the sum of which are in the player's inventory
+#                       must be 1.0+ to allow logical completion
+#
+# Each Star's logical requirements imply any previous requirements
+#
+level_logic = {
```

Comment:
This could be reduced to sections with actual data using things like collections.defaultdict, if you wish,

Reply:
I think I will keep as is for now